warehouse/oidc/forms/activestate.py

```diff
@@ -0,0 +1,210 @@
+    __params__ = ["organization", "project", "actor"]
+
+    organization = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(
+                message=_("Specify ActiveState organization name"),
+            ),
+            wtforms.validators.Regexp(
+                _VALID_ORG_URL_NAME_AND_ACTOR_NAME,
+                message=_("Invalid ActiveState organization name"),
+            ),
+            _no_double_dashes,
+            _no_leading_or_trailing_dashes,
+        ]
+    )
+
+    project = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(
+                message=_("Specify ActiveState project name")
+            ),
+            wtforms.validators.Regexp(
+                _VALID_PROJECT_NAME,
+                message=_("Invalid ActiveState project name"),
+            ),
+            wtforms.validators.Regexp(
+                _VALID_PROJECT_NAME,
+                message=_("Invalid ActiveState project name"),
+            ),
+            _no_double_dashes,
+            _no_leading_or_trailing_dashes,
+        ]
+    )
+
+    actor = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(
+                message=("Specify the ActiveState actor username")
+            ),
+            wtforms.validators.Regexp(
+                _VALID_ORG_URL_NAME_AND_ACTOR_NAME,
+                message=("Invalid ActiveState username"),
+            ),
+            _no_double_dashes,
+            _no_leading_or_trailing_dashes,
+        ]
+    )
+
+    def _lookup_organization(self, org_url_name: str) -> None:
+        """Make gql API call to the ActiveState API to check if the organization
+        exists"""
+
+        def process_org_response(response: GqlResponse) -> None:
+            data = response.get("data")
+            if data and not data.get("organizations"):
+                raise wtforms.validators.ValidationError(
+                    _("ActiveState organization not found")
+                )
+
+        return _activestate_gql_api_call(
+            _GRAPHQL_GET_ORGANIZATION, {"orgname": org_url_name}, process_org_response
+        )
+
+    def validate_organization(self, field):
+        self._lookup_organization(field.data)
+
+    def _lookup_actor(self, actor: str) -> UserResponse:
+        """Make gql API call to the ActiveState API to check if the actor/username
+        exists and return the associated user id"""
+
+        def process_actor_response(response: GqlResponse) -> UserResponse:
+            users = response.get("data", {}).get("users", [])
+            if users:
+                return users[0]
+            else:
+                raise wtforms.validators.ValidationError(
+                    _("ActiveState actor not found")
+                )
+
+        return _activestate_gql_api_call(
+            _GRAPHQL_GET_ACTOR, {"username": actor}, process_actor_response
+        )
+
+    def validate_actor(self, field):
```

Comment:
I was reading more about wtforms validators, and I _think_ that if there's a named validator function, it fires **before** the validators on the field, so `validate_actor()` would perform API calls before the other validators like regex/dashes are in play.

It's definitely worth re-verifying - since if our validators operate in that order, then we're performing an API call regardless. I hope I'm wrong! If that _is_ true, then it's not only this form - so I don't think it's immediately actionable, but if you can test it, that'd be great.

Reply:
If you like the idea here https://github.com/pypi/warehouse/pull/15168#discussion_r1480490211. I would do that change for this validator as well.  We really just need a "you entered the wrong text" kind of feedback here.